kvserver: move SideloadStorage to a package

pav-kv opened the pull request:
Fixes #91899
Release note: None

pav-kv committed:
kvserver: move syncing_write.go to kvserverbase

This file is used both by log storage (storing AddSST commands in sideloaded
storage) and state storage (applying AddSST). Moving it to kvserverbase to
avoid circular dependencies when factoring out the log storage.

Release note: None

pav-kv committed:
kvserver: move SideloadStorage to a package

This commit moves and exports the SideloadStorage interface and its
implementation to the new logstore package. The package will accumulate types
forming the separated log storage.

Only the unit tests which did not depend on Replica are moved to the new
package. Tests which depended on Replica stay in kvserver for now, with a TODO
to untangle them from Replica and move away. They might stay too because they
test integration between kvserver and logstore.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91900)
<!-- Reviewable:end -->

pavelkalinnikov commented:
Pending #91898 merge, and figuring out where to place the `syncing_write.go` file.

pavelkalinnikov commented:
Ready for review (only need to look at the last 2 commits).

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload.go:
```diff
@@ -72,7 +72,7 @@ type SideloadStorage interface {
 // The provided slice is not modified, though the returned slice may be backed
 // in parts or entirely by the same memory.
-func maybeSideloadEntries(
+func MaybeSideloadEntries(
```
All these will be unexported again when more things move to `logstore`.

pavelkalinnikov commented:
I don't understand the CI failure. Does it look like another circular dependency?

pav-kv committed:
kvserver: reduce import scope to storage/fs

Release note: None

pav-kv committed:
kvserver: move helper to where it's used

Move newOnDiskEngine helper to replica_sst_snapshot_storage_test.go, the only
place where it's used.

Release note: None

tbg commented on pkg/kv/kvserver/logstore/sideload.go:
```diff
@@ -23,7 +23,7 @@ import (
 )
 
-var errSideloadedFileNotFound = errors.New("sideloaded file not found")
+var ErrSideloadedFileNotFound = errors.New("sideloaded file not found")
```
Doesn't this need a comment or linters will complain?

tbg replied:
Huh, apparently not.

pavelkalinnikov replied:
Yeah, maybe this lint is disabled. I'm unexporting this though, it's used only inside the package. Will now revisit/fix the exporting/commenting in this PR, and then merge.

pav-kv committed:
logstore: comment exported surface

Release note: None

pavelkalinnikov commented:
bors r=tbg
